File: src/main/scala/com/github/shadowsocks/plugin/kcptun/HelpCallback.scala
package com.github.shadowsocks.plugin.kcptun

import com.github.shadowsocks.plugin.PluginOptions

import scala.io.Source

/**
  * @author Mygod
  */
class HelpCallback extends com.github.shadowsocks.plugin.HelpCallback {
  override def produceHelpMessage(options: PluginOptions): CharSequence = {
    val process = new ProcessBuilder(getApplicationInfo.nativeLibraryDir + "/libkcptun.so", "--help")
      .redirectErrorStream(true)
      .start()
    process.waitFor()
    Source.fromInputStream(process.getInputStream())
      .getLines()
      .dropWhile(line => line != "GLOBAL OPTIONS:")
      .drop(1)
      .takeWhile(line => line.length() > 3)
      .filter(line =>
        !line.startsWith("   --localaddr ") &&
        !line.startsWith("   --remoteaddr ") &&
        !line.startsWith("   -c ") &&
        !line.startsWith("   -V ") &&
        !line.startsWith("   --fast-open ") &&
        !line.startsWith("   --help,") &&
        !line.startsWith("   --version,"))
      .mkString("\n")
      .replaceAll(" {2,}", "\n")
      .replaceAll("--", "")
      .replaceAll(" value", "=value;")
      .substring(1) // remove 1st \n
  }
}

File: app/src/main/java/com/github/shadowsocks/plugin/kcptun/HelpCallback.kt
package com.github.shadowsocks.plugin.kcptun

import com.github.shadowsocks.plugin.PluginOptions

class HelpCallback : com.github.shadowsocks.plugin.HelpCallback() {
    override fun produceHelpMessage(options: PluginOptions): CharSequence =
            ProcessBuilder(applicationInfo.nativeLibraryDir + "/libkcptun.so", "--help")
                    .redirectErrorStream(true)
                    .start()
                    .inputStream.bufferedReader().useLines {
                it.dropWhile { it != "GLOBAL OPTIONS:" }
                        .drop(1)
                        .takeWhile { it.length > 3 }
                        .filter {
                            !it.startsWith("   --localaddr ") &&
                                    !it.startsWith("   --remoteaddr ") &&
                                    !it.startsWith("   --log ") &&
                                    !it.startsWith("   --quiet ") &&
                                    !it.startsWith("   -c ") &&
                                    !it.startsWith("   -V ") &&
                                    !it.startsWith("   --fast-open ") &&
                                    !it.startsWith("   --help,") &&
                                    !it.startsWith("   --version,")
                        }
                        .joinToString("\n")
                        .replace(Regex(" {2,}"), "\n")
                        .replace("--", "")
                        .replace(" value", "=value")
                        .substring(1)   // remove 1st \n
            }
}
